clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -507,45 +508,81 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
     if (type_is_instr(memref.instr.type) ||
         memref.instr.type == TRACE_TYPE_PREFETCH_INSTR ||
         memref.instr.type == TRACE_TYPE_INSTR_NO_FETCH) {
+        per_shard_t::instr_info_t cur_instr_info = {};
+        const bool expect_encoding =
+            TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->file_type_);
         // We'd prefer to report this error at the syscall instr but it is easier
         // to wait until here:
         report_if_false(shard,
                         !TESTANY(OFFLINE_FILE_TYPE_SYSCALL_NUMBERS, shard->file_type_) ||
                             !shard->expect_syscall_marker_,
                         "Syscall marker missing after syscall instruction");
-        bool expect_encoding = TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->file_type_);
-        std::unique_ptr<instr_autoclean_t> cur_instr_decoded = nullptr;
+
         if (expect_encoding) {
-            cur_instr_decoded.reset(new instr_autoclean_t(GLOBAL_DCONTEXT));
-            // TODO i#6006: cache the decoding results, and avoid heap with
-            // instr_noalloc_t instead of instr_autoclean_t.
-            app_pc next_pc = decode_from_copy(
-                GLOBAL_DCONTEXT, const_cast<app_pc>(memref.instr.encoding),
-                reinterpret_cast<app_pc>(memref.instr.addr), cur_instr_decoded->data);
-            if (next_pc == nullptr) {
-                cur_instr_decoded.reset(nullptr);
+            const app_pc trace_pc = reinterpret_cast<app_pc>(memref.instr.addr);
+            // Clear cache entry for new encodings.
+            if (memref.instr.encoding_is_new) {
+                shard->decode_cache_.erase(trace_pc);
+            }
+            auto cached = shard->decode_cache_.find(trace_pc);
+            if (cached != shard->decode_cache_.end()) {
+                cur_instr_info.decoding = cached->second;
+            } else {
+                instr_noalloc_t noalloc;
+                instr_noalloc_init(drcontext_, &noalloc);
+                instr_t *noalloc_instr = instr_from_noalloc(&noalloc);
+                app_pc next_pc = decode_from_copy(
+                    drcontext_, const_cast<app_pc>(memref.instr.encoding),
+                    reinterpret_cast<app_pc>(memref.instr.addr), noalloc_instr);
+                if (next_pc == nullptr) {
+                    noalloc_instr = nullptr;
+                }
+                // Add decoding attributes to cur_instr_info.
+                if (noalloc_instr != nullptr) {
```

Comment:
I meant: remove the `if (noalloc_instr != nullptr)`.  Since `noalloc_instr` is non-null by construction before the decode_from_copy call, and decode_from_copy cannot set it to null, it seems weird to convert the return value from decode_from_copy into it being null.  We still need to check that the decoding succeeded.  I was suggesting to put all the valid decoding code inside `if (next_pc != nullptr)`.  (Certainly the new code is incorrect as it will mark it valid even if the decoding fails.)